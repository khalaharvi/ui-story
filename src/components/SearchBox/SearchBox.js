import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { connectStats, connectSearchBox } from 'react-instantsearch/connectors';

import { Reset as ResetIcon } from 'icons';
import { createClassNames } from 'utils';

const cx = createClassNames('search-box');

export class SearchBox extends Component {
  static propTypes = {
    currentRefinement: PropTypes.string,
    className: PropTypes.string,
    refine: PropTypes.func.isRequired,
    nbHits: PropTypes.number,

    onSubmit: PropTypes.func,
    onReset: PropTypes.func,
    onChange: PropTypes.func,
    placeholder: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),

    isSearchStalled: PropTypes.bool
  };

  static defaultProps = {
    currentRefinement: '',
    className: '',
    isSearchStalled: false
  };

  onSubmit = (e) => {
    e.preventDefault();
    e.stopPropagation();
    this.input.blur();

    return false;
  };

  onChange = (e) => {
    const { refine, onChange } = this.props;

    refine(e.target.value);

    if (onChange) {
      onChange(e);
    }
  };

  onReset = (e) => {
    const { refine, onReset } = this.props;

    refine('');
    this.input.focus();

    if (onReset) {
      onReset(e);
    }
  };

  getQuery = () => this.props.currentRefinement;
  getPlaceholder = () => {
    const { placeholder, nbHits } = this.props;

    if (placeholder) {
      return (typeof placeholder === 'function') ? placeholder(nbHits) : placeholder;
    }

    return null;
  }

  render() {
    const {
      className,
      isSearchStalled
    } = this.props;

    const query = this.getQuery();
    const placeholder = this.getPlaceholder();
    const searchInputEvents = Object.keys(this.props).reduce((props, prop) => {
      if (
        ['onsubmit', 'onreset', 'onchange'].indexOf(prop.toLowerCase()) === -1 &&
        prop.indexOf('on') === 0
      ) {
        return { ...props, [prop]: this.props[prop] };
      }

      return props;
    }, {});

    return (
      <div className={classNames(cx(''), className)}>
        <form
          noValidate
          onSubmit={this.props.onSubmit ? this.props.onSubmit : this.onSubmit}
          onReset={this.onReset}
          className={cx('form', isSearchStalled && 'form--stalledSearch')}
          action=""
          role="search"
        >
          <div className={cx('input-container')}>
            <input
              ref={(input) => { this.input = input; }}
              type="search"
              placeholder={placeholder}
              autoComplete="off"
              autoCorrect="off"
              autoCapitalize="off"
              spellCheck="false"
              value={query}
              onChange={this.onChange}
              {...searchInputEvents}
              className={cx('input')}
            />
            <button hidden={!query || isSearchStalled} type="reset" className={cx('input-button')}>
              <ResetIcon className={cx('input-icon')} />
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default connectStats(connectSearchBox(SearchBox));
