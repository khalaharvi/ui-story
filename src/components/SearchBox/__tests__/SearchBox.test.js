import React from 'react';
import { mount, shallow } from 'enzyme';

import { SearchBox } from '../SearchBox';

describe('<SearchBox />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<SearchBox refine={() => null} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with custom classname', () => {
    const wrapper = shallow(<SearchBox refine={() => null} className="my-custom-classname" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('treats its query prop as its input value', () => {
    const wrapper = shallow(
      <SearchBox refine={() => null} currentRefinement="QUERY1" />
    );

    expect(wrapper).toMatchSnapshot();

    wrapper.update(
      <SearchBox refine={() => null} currentRefinement="QUERY2" />
    );

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('should accept `onXXX` events', () => {
    const onSubmit = jest.fn();
    const onReset = jest.fn();

    const inputEventsList = [
      'onChange',
      'onFocus',
      'onBlur',
      'onSelect',
      'onKeyDown',
      'onKeyPress'
    ];

    const inputProps = inputEventsList.reduce(
      (props, prop) => ({ ...props, [prop]: jest.fn() }),
      {}
    );

    const wrapper = mount(
      <SearchBox
        refine={() => null}
        onSubmit={onSubmit}
        onReset={onReset}
        {...inputProps}
      />
    );

    // simulate form events `onReset` && `onSubmit`
    wrapper.find('form').simulate('submit');
    expect(onSubmit).toBeCalled();

    wrapper.find('form').simulate('reset');
    expect(onReset).toBeCalled();

    // simulate input search events
    inputEventsList.forEach((eventName) => {
      wrapper
        .find('input')
        .simulate(eventName.replace(/^on/, '').toLowerCase());

      expect(inputProps[eventName]).toBeCalled();
    });

    wrapper.unmount();
  });

  it('should clear the query when the form is reset', () => {
    const refine = jest.fn();

    const wrapper = mount(<SearchBox refine={refine} />);

    // Simulate the ref
    wrapper.instance().input = { focus: jest.fn() };

    wrapper.find('form').simulate('reset');

    expect(refine).toHaveBeenCalledWith('');
    expect(wrapper.instance().input.focus).toHaveBeenCalled();

    wrapper.unmount();
  });
});
