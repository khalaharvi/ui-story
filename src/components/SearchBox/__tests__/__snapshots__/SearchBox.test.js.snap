// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> renders correctly 1`] = `
<div
  className="dh-search-box"
>
  <form
    action=""
    className="dh-search-box-form"
    noValidate={true}
    onReset={[Function]}
    onSubmit={[Function]}
    role="search"
  >
    <div
      className="dh-search-box-input-container"
    >
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        className="dh-search-box-input"
        onChange={[Function]}
        placeholder={null}
        spellCheck="false"
        type="search"
        value=""
      />
      <button
        className="dh-search-box-input-button"
        hidden={true}
        type="reset"
      >
        <Reset
          className="dh-search-box-input-icon"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`<SearchBox /> renders correctly with custom classname 1`] = `
<div
  className="dh-search-box my-custom-classname"
>
  <form
    action=""
    className="dh-search-box-form"
    noValidate={true}
    onReset={[Function]}
    onSubmit={[Function]}
    role="search"
  >
    <div
      className="dh-search-box-input-container"
    >
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        className="dh-search-box-input"
        onChange={[Function]}
        placeholder={null}
        spellCheck="false"
        type="search"
        value=""
      />
      <button
        className="dh-search-box-input-button"
        hidden={true}
        type="reset"
      >
        <Reset
          className="dh-search-box-input-icon"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`<SearchBox /> treats its query prop as its input value 1`] = `
<div
  className="dh-search-box"
>
  <form
    action=""
    className="dh-search-box-form"
    noValidate={true}
    onReset={[Function]}
    onSubmit={[Function]}
    role="search"
  >
    <div
      className="dh-search-box-input-container"
    >
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        className="dh-search-box-input"
        onChange={[Function]}
        placeholder={null}
        spellCheck="false"
        type="search"
        value="QUERY1"
      />
      <button
        className="dh-search-box-input-button"
        hidden={false}
        type="reset"
      >
        <Reset
          className="dh-search-box-input-icon"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`<SearchBox /> treats its query prop as its input value 2`] = `
<div
  className="dh-search-box"
>
  <form
    action=""
    className="dh-search-box-form"
    noValidate={true}
    onReset={[Function]}
    onSubmit={[Function]}
    role="search"
  >
    <div
      className="dh-search-box-input-container"
    >
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        className="dh-search-box-input"
        onChange={[Function]}
        placeholder={null}
        spellCheck="false"
        type="search"
        value="QUERY1"
      />
      <button
        className="dh-search-box-input-button"
        hidden={false}
        type="reset"
      >
        <Reset
          className="dh-search-box-input-icon"
        />
      </button>
    </div>
  </form>
</div>
`;
